Internal compiler error anonymous bound region BrAnon(0) in binding but not trait ref

The following code causes an error

```rust
impl<'b, B: Ty<'b>, I: SIterator, F> SIterator for FilterMap<I, F> where
    F: FnMut(<I as Ty>::V) -> Option<<B as Ty>::V>
{}

pub trait SIterator: for<'a> Ty<'a> {}

pub trait Ty<'a> {
    type V;
}

```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e175d6557baecbf033a125d9276bd385))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0465]: multiple rlib candidates for `debug_unreachable` found
  |
  = note: candidate #1: /playground/target/debug/deps/libdebug_unreachable-53cbbc8451f8a571.rlib
  = note: candidate #2: /playground/target/debug/deps/libdebug_unreachable-9f86ee3a7179ede7.rlib

error[E0412]: cannot find type `FilterMap` in this scope
 --> src/lib.rs:1:52
  |
1 | impl<'b, B: Ty<'b>, I: SIterator, F> SIterator for FilterMap<I, F> where
  |                                                    ^^^^^^^^^ not found in this scope
help: possible candidates are found in other modules, you can import them into scope
  |
1 | use core::iter::FilterMap;
  |
1 | use futures::stream::FilterMap;
  |
1 | use itertools::__std_iter::FilterMap;
  |
1 | use rayon::iter::FilterMap;
  |
and 2 other candidates

error: internal compiler error: src/librustc_typeck/astconv.rs:1097: anonymous bound region BrAnon(0) in binding but not trait ref
 --> src/lib.rs:2:31
  |
2 |     F: FnMut(<I as Ty>::V) -> Option<<B as Ty>::V>
  |                               ^^^^^^^^^^^^^^^^^^^^

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:578:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0412`.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.37.0-nightly (b25ee6449 2019-06-17) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

If it helps, this slightly modified snippet produces a very odd error:

```rust
struct S {}

trait T<'a>{
    type A;
}

impl T<'_> for S {
    type A = i32;
}

fn foo<'a>(x: impl Fn(<S as T>::A) -> <S as T<'a>>::A) {
    x(2i32);
}

fn main() {
    foo(|x| x);
}
```

```
error[E0308]: mismatched types
  --> src/main.rs:12:7
   |
12 |     x(2i32);
   |       ^^^^ expected associated type, found `i32`
   |
   = note: expected associated type `<S as T<'_>>::A`
                         found type `i32`
   = note: consider constraining the associated type `<S as T<'_>>::A` to `i32` or calling a method that returns `<S as T<'_>>::A`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
```

In that code `<S as T<'_>>::A` is `i32`, so it seems like something isn't being fully expanded somewhere?

--- a/src/librustc_typeck/astconv.rs
+++ b/src/librustc_typeck/astconv.rs
@@ -1485,28 +1485,33 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 debug!("late_bound_in_ty = {:?}", late_bound_in_ty);
                 for br in late_bound_in_ty.difference(&late_bound_in_trait_ref) {
                     let br_name = match *br {
-                        ty::BrNamed(_, name) => name,
-                        _ => {
-                            span_bug!(
-                                binding.span,
-                                "anonymous bound region {:?} in binding but not trait ref",
-                                br
-                            );
-                        }
+                        ty::BrNamed(_, name) => format!("lifetime `{}`", name),
+                        _ => "an anonymous lifetime".to_string(),
                     };
                     // FIXME: point at the type params that don't have appropriate lifetimes:
                     // struct S1<F: for<'a> Fn(&i32, &i32) -> &'a i32>(F);
                     //                         ----  ----     ^^^^^^^
-                    struct_span_err!(
+                    let mut err = struct_span_err!(
                         tcx.sess,
                         binding.span,
                         E0582,
-                        "binding for associated type `{}` references lifetime `{}`, \
+                        "binding for associated type `{}` references {}, \
                          which does not appear in the trait input types",
                         binding.item_name,
                         br_name
-                    )
-                    .emit();
+                    );
+
+                    if let ty::BrAnon(_) = *br {
+                        // The only way for an anonymous lifetime to wind up
+                        // in the return type but **also** be unconstrained is
+                        // if it only appears in "associated types" in the
+                        // input. See #62200 for an example. In this case,
+                        // though we can easily give a hint that ought to be
+                        // relevant.
+                        err.note("lifetimes appearing in an associated type are not considered constrained");
+                    }
+
+                    err.emit();
                 }
             }
         }

--- a/src/test/ui/associated-types/issue-62200.rs
+++ b/src/test/ui/associated-types/issue-62200.rs
@@ -1,0 +1,14 @@
+struct S {}
+
+trait T<'a> {
+    type A;
+}
+
+impl T<'_> for S {
+    type A = u32;
+}
+
+fn foo(x: impl Fn(<S as T<'_>>::A) -> <S as T<'_>>::A) {}
+//~^ ERROR binding for associated type `Output` references an anonymous lifetime
+
+fn main() {}

--- a/src/test/ui/associated-types/issue-62200.rs
+++ b/src/test/ui/associated-types/issue-62200.rs
@@ -10,5 +10,6 @@ impl T<'_> for S {
 
 fn foo(x: impl Fn(<S as T<'_>>::A) -> <S as T<'_>>::A) {}
 //~^ ERROR binding for associated type `Output` references an anonymous lifetime
+//~^^ NOTE lifetimes appearing in an associated type are not considered constrained
 
 fn main() {}

--- a/src/test/ui/associated-types/issue-62200.stderr
+++ b/src/test/ui/associated-types/issue-62200.stderr
@@ -1,0 +1,11 @@
+error[E0582]: binding for associated type `Output` references an anonymous lifetime, which does not appear in the trait input types
+  --> $DIR/issue-62200.rs:11:39
+   |
+LL | fn foo(x: impl Fn(<S as T<'_>>::A) -> <S as T<'_>>::A) {}
+   |                                       ^^^^^^^^^^^^^^^
+   |
+   = note: lifetimes appearing in an associated type are not considered constrained
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0582`.

